opm/simulators/flow/partitionCells.cpp

```diff
@@ -29,46 +41,141 @@
 #include <stdexcept>
 #include <string>
 
-namespace Opm
+namespace {
+
+std::pair<std::vector<int>, int>
+countDomains(std::vector<int> partition_vector)
 {
+    auto maxPos = std::max_element(partition_vector.begin(),
+                                   partition_vector.end());
+
+    const auto num_domains = (maxPos == partition_vector.end())
+        ? 0 : *maxPos + 1;
+
+    return { std::move(partition_vector), num_domains };
+}
 
+template <typename, class = void>
+struct HasZoltanPartitioning : public std::false_type {};
 
-    // Read from file, containing one number per cell.
-    std::pair<std::vector<int>, int> partitionCellsFromFile(const std::string& filename, const int num_cells)
-    {
-        // Read file into single vector.
-        std::ifstream is(filename);
-        const std::vector<int> cellpart{std::istream_iterator<int>(is), std::istream_iterator<int>()};
-        if (cellpart.size() != size_t(num_cells)) {
-            auto msg = fmt::format("Partition file contains {} entries, but there are {} cells.",
-                                   cellpart.size(), num_cells);
-            throw std::runtime_error(msg);
+template <typename GridType>
+struct HasZoltanPartitioning<
+    GridType,
+    std::void_t<decltype(std::declval<GridType>().zoltanPartitionWithoutScatter
```

Comment:
You'll strictly speaking need `<type_traits>` for `std::void_t`.

Reply:
> indeed, fixed.

Cool.  I did overlook one other issue, though.  The way we're using `zoltanPartitionWithoutScatter()`, we'll have to be a little be more careful here.  We're passing `GridType` directly, but the parameter type is `const GridType&`.  The two parameter types must match lest we get build failures that are hard to analyse in case the `GridType` only has a mutating `zoltanPartitionWithoutScatter()` member function.  The easiest fix is probably to write
```C++
std::void_t<decltype(std::declval<const GridType&>().zoltanPartitionWithoutScatter
```
instead.

Reply:
sure, mutable "safety" added.